### PR TITLE
[JENKINS-73224] Do not try to create a pod which already exists

### PR DESCRIPTION
**Pull request by amuniz**
See [JENKINS-73224](https://issues.jenkins.io/browse/JENKINS-73224) for details.

Easier to review hiding whitespace changes: https://github.com/jenkinsci/kubernetes-plugin/pull/1561/files?w=1

### Testing done

Automated test added.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java [hunk 1]
```diff
@@ -12,6 +20,45 @@ public KubernetesPipelineRJRTest() throws UnknownHostException {
...
+                String nodeName =
+                        client.nodes().list().getItems().get(0).getMetadata().getName();
+                client.nodes().withName(nodeName).edit(n -> new NodeBuilder(n)
+                        .editMetadata()
+                        .removeFromLabels("disktype")
+                        .endMetadata()
+                        .build());
+            }
+        }
+    }
```
Comment (amuniz) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java:63]: Without the fix this test fails with the log in the linked ticket.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/AbstractKubernetesPipelineRJRTest.java [hunk 2]
```diff
@@ -52,7 +47,5 @@ public static void isKubernetesConfigured() throws Exception {
     public void setUp() throws Throwable {
         rjr.startJenkins();
         rjr.runRemotely(setup);
-        runId = rjr.runRemotely(new CreateWorkflowJobThenScheduleRun(
-                KubernetesTestUtil.loadPipelineScript(getClass(), name.getMethodName() + ".groovy")));
```
Comment (amuniz) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/AbstractKubernetesPipelineRJRTest.java:56]: Moved down to tests as not all of them need a job created and triggered as setup.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 3]
```diff
@@ -129,51 +129,71 @@ public synchronized void launch(SlaveComputer computer, TaskListener listener) {
...
-                        node.getRunListener()
-                                .getLogger()
-                                .printf(
-                                        "WARNING: Unable to create pod: %s %s/%s because kubernetes resource quota update conflict. %n%s%nRetrying...%n%n",
-                                        cloudName, namespace, pod.getMetadata().getName(), e.getMessage());
+            // if the controller was interrupted after creating the pod but before it connected back, then
+            // the pod might already exist and the creating logic must be skipped.
+            Pod existingPod =
+                    client.pods().inNamespace(namespace).withName(podName).get();
+            if (existingPod == null) {
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:136]: Everything inside this `if` branch is untouched (except formatting, which I was forced to do by Spotless)

**Review by Vlatombe — APPROVED** [review]
LGTM

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 4]
```diff
@@ -129,51 +129,71 @@ public synchronized void launch(SlaveComputer computer, TaskListener listener) {
...
+            } else {
+                LOGGER.log(INFO, () -> "Pod already exists: " + cloudName + " " + namespace + "/" + podName);
+                listener.getLogger().printf("Pod already exists: %s %s/%s%n", cloudName, namespace, podName);
             }
-            LOGGER.log(INFO, () -> "Created Pod: " + cloudName + " " + namespace + "/" + podName);
-            listener.getLogger().printf("Created Pod: %s %s/%s%n", cloudName, namespace, podName);
-            Metrics.metricRegistry().counter(MetricNames.PODS_CREATED).inc();
-
-            node.getRunListener().getLogger().printf("Created Pod: %s %s/%s%n", cloudName, namespace, podName);
             kubernetesComputer.setLaunching(true);
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:197]: Could we fix this field please, so it is not persisting an `AtomicBoolean`?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 5]
```diff
@@ -68,7 +67,7 @@ public class KubernetesLauncher extends JNLPLauncher {
 
     private static final Logger LOGGER = Logger.getLogger(KubernetesLauncher.class.getName());
 
-    private final AtomicBoolean launched = new AtomicBoolean(false);
+    private volatile boolean launched = false;
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:70]: FTR: this is not directly related or needed for the goal of this PR, but it was requested by @jglick in a review.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 6]
```diff
@@ -129,51 +128,71 @@ public synchronized void launch(SlaveComputer computer, TaskListener listener) {
...
                 }
-                throw e;
+                LOGGER.log(INFO, () -> "Created Pod: " + cloudName + " " + namespace + "/" + podName);
+                listener.getLogger().printf("Created Pod: %s %s/%s%n", cloudName, namespace, podName);
+                Metrics.metricRegistry().counter(MetricNames.PODS_CREATED).inc();
+
+                node.getRunListener().getLogger().printf("Created Pod: %s %s/%s%n", cloudName, namespace, podName);
+            } else {
+                LOGGER.log(INFO, () -> "Pod already exists: " + cloudName + " " + namespace + "/" + podName);
+                listener.getLogger().printf("Pod already exists: %s %s/%s%n", cloudName, namespace, podName);
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:194]: So how does this actually differ from https://github.com/jenkinsci/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L106-L110 and do we even still need the `launching` field if we have this logic?

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java [hunk 7]
```diff
@@ -12,6 +19,44 @@ public KubernetesPipelineRJRTest() throws UnknownHostException {
...
+    public void restartDuringPodLaunch() throws Throwable {
+        // try to run something on a pod which is not schedulable (disktype=special)
+        RunId build = rjr.runRemotely(new CreateWorkflowJobThenScheduleTask(
+                KubernetesTestUtil.loadPipelineScript(getClass(), name.getMethodName() + ".groovy")));
+        // the pod is created, but not connected yet
+        rjr.runRemotely(new AssertBuildLogMessage("Created Pod", build));
+        // restart
+        rjr.stopJenkins();
+        rjr.startJenkins();
+        // update k8s to make a node suitable to schedule (add disktype=special to the node)
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java:36]: A little scary to have the test even temporarily patch the `Node`! Would be safer to do something to the namespace like a quota :thinking: Fine enough for Kind/CI, could just become a problem if we try to run tests on a permanent cluster.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package org.csanchez.jenkins.plugins.kubernetes.pipeline.steps;
+
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.RealJenkinsRule;
+
+public class AssertBuildLogMessage implements RealJenkinsRule.Step {
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java:8]: BTW you can also make these things `static` methods in some util class; a bit less boilerplate.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 5 above)
Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:70]: (but consider https://github.com/jenkinsci/kubernetes-plugin/pull/1561#discussion_r1617565856)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 6 above)
Comment (amuniz) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:194]: That is another face of the same behaviour, just under a slightly different timing (a lot narrower than the one being fixed in this PR).

This PR catches the case there the controller is restarted between the [pod creating call](https://github.com/jenkinsci/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L138) and the end of the [waiting for loop](https://github.com/amuniz/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L279). While the existing `launched` check would only catch a restart right after the [node save](https://github.com/amuniz/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L293) and the end of the method, which is quite unlikely to happen, but possible I guess.

Comment (amuniz) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:194]: >  do we even still need the launching field if we have this logic?

Not sure. https://github.com/amuniz/kubernetes-plugin/blob/f10083a42e708a6bff990c6ccb306ce618b8bccd/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesComputer.java#L201 is never used from this plugin sources, but it's public and it could be used somewhere else, so I would stay on the safe side and keep it.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java (hunk 8 above)
Comment (amuniz) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java:8]: Yeah, but less reusable then...

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java (hunk 7 above)
Comment (amuniz) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java:36]: This was the easiest way I found, and with the `finally` it's granted to be back to the original state.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 6 above)
Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:194]: > is never used from this plugin sources, but it's public and it could be used somewhere else

Well, if it is not used it can be deleted. (I usually just do a GH org-wide search.)

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java (hunk 7 above)
Comment (jglick) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineRJRTest.java:36]: So long as the JVM terminates normally, yes.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java (hunk 8 above)
Comment (jglick) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java:8]: How is that less reusable?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 6 above)
Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:194]: Then there is stuff like https://github.com/amuniz/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L199-L200 which it would not make sense to do twice, but there is no current defense against race conditions. It seems there are deeper bugs waiting here.

**Review by Vlatombe — APPROVED** [review]

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java (hunk 8 above)
Comment (amuniz) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java:8]: When writing a new test it seems natural to go to the `steps` package and look for generic steps to use in the new test, however inspecting other test classes looking for static methods seems less discoverable to me at least. That will likely end up with re-writing the same (or very similar) static methods in multiple test classes.

Comment (jglick) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/steps/AssertBuildLogMessage.java:8]: > inspecting other test classes looking for static methods

I meant reusable static methods in a utility class in the `steps` package.
